feature/CATC_144-connect-frontend-with-backend

sabellius opened the pull request:
## 🎯 Description
This PR connects the Trello clone frontend with the backend server, replacing local storage with REST API calls for real data persistence and multi-user functionality.

## 🔧 Changes
- Connected both auth and board services (remote versions) with the backend.

## 📝 Notes
- Made some fixes to bugs i found in the backend/frontend along the way.
- Some function are still commented out and wont work in at all as they aren't implemented in the server yet.
- Many of the implementations are breaking changes  as some function signatures changed and using the local service will not be possible anymore.
- Some state changes in the frontend were made to work with different responses from the server from what we worked with locally. for example when we move a list, we don't want to query all the board lists all over, and instead we only return the moved list, with which hte frontend will know whether to filter it out or to resort the lists array (no need to send data the frontend already has, and this also is consistent with REST way of doing things).
- The following functoin currently wont work, as most of them are not yet implemented in the backend. Will work on them ASAP.

sabellius committed:
feat(auth): authenticate user using the real server

sabellius committed:
feat(create-card)!: create cards using the remote server

BREAKING CHANGE: the return value of the function addCard has changed to
return the created card directly instead of the entire list.
The local-service will now stop working.

sabellius committed:
feat(edit-card)!: edit cards using the remote server

BREAKING CHANGE: The editCard function now return only the updated card,
as the listId is already inside the card data itself. local-service
will stop working. Reducer updated accordingly.

sabellius committed:
feat(delete-card): delete cards using the remote server

sabellius committed:
refactor: use a utility function to make it consistent with the rest of the file

sabellius committed:
fix: fix list creation

- Create a default position (at the end) when creating a new list to
avoid required field error.
- Destructure only accepted fields when creating a new list.

sabellius committed:
feat(create-list): create lists using the backend server

sabellius committed:
fix: temporariy fix to allow board members to edit board

sabellius committed:
feat(create-label): create board labels using the remote server

sabellius committed:
feat(edit-label): edit board labels using the remote server

sabellius committed:
fix(delete-label): return nothing as we don't need any data

sabellius committed:
feat(delete-label): delete labels using the remote server

sabellius committed:
feat(board-previews): get previews from the remote server

sabellius committed:
feat(board-list-previews): get previews from the remote server

sabellius committed:
feat(copy-card): copy cards using the remote server

sabellius committed:
feat(move-card): move cards using the remote server

sabellius committed:
fix: get list and board options from the server

sabellius committed:
feat(move-list): move lists using the remote server

sabellius committed:
fix: import the correct service dynamically

sabellius committed:
feat: use modified getEmptyCard with correct default values

sabellius committed:
fix: return correct value

sabellius committed:
fix: return correct data

sabellius committed:
fix: return correct data

sabellius committed:
feat(copy-list): copy lists using the remote service

snoxlax commented on frontend/src/store/reducers/board-reducer.js:
```diff
@@ -109,11 +109,22 @@ const handlers = {
   }),
   ...createAsyncHandlers(MOVE_LIST, MOVE_LIST.KEY),
   [MOVE_LIST.SUCCESS]: (state, action) => {
+    let newLists;
+    if (state.board._id !== action.payload.boardId) {
+      newLists = state.board.lists.filter(
+        list => list._id !== action.payload._id
+      );
+    } else {
+      newLists = state.board.lists.map(list =>
+        list._id === action.payload._id ? { ...list, ...action.payload } : list
+      );
+      newLists = sortByPosition(newLists);
+    }
```
can we put this inside the service instead of here?

sabellius replied:
We could, but I think the purpose of the service is to abstract the interaction with the back-end, do we really want to mix the UI state manipulation in there as well?

snoxlax commented on frontend/src/store/reducers/board-reducer.js:
```diff
@@ -46,7 +46,7 @@ const handlers = {
     board: {
       ...state.board,
-      lists: [...state.board.lists, action.payload],
+      lists: sortByPosition([...state.board.lists, action.payload]),
```
Is it possible to move this into the service? I think that would be cleaner

sabellius replied:
Again, this probably should go into the service because it's not related to the logic of the service, the services' purpose is to fetch or send data to the server and receive whatever the server responds with. For example in this case, we add a list, the service send some data to the server, and receives a newly created list object. The service doesn't and shouldn't know about other things and contexts such as lists how how the returned data should be manipulated in the UI. If the reducer cases seem too long to you, perhaps a better solution would be to extract some of this logic into store utility/helper functions, but in terms of boundaries, I don't think we should mix this we the board service. What do you think?

snoxlax commented on frontend/src/store/reducers/board-reducer.js:
```diff
@@ -133,7 +144,12 @@ const handlers = {
+          ? list
+          : {
+              ...list,
+              cards: sortByPosition([...list.cards, action.payload]),
```
and this

sabellius committed:
fix: prevent duplicate cards when moving within same list

- Changed MOVE_CARD.SUCCESS reducer to filter card from all lists first
- Then add the card only to the target list
- This fixes the duplicate card bug when moving cards within the same list

snoxlax reviewed (approved):
Looks good!
